redismaint/rmaint.go

```diff
@@ -0,0 +1,131 @@
+package redismaint
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/gomodule/redigo/redis"
+)
+
+type (
+	SendRequestWithPubSub func(ctx context.Context, url string, action string, payload []byte, header map[string]string, key string) (int, []byte, error)
+)
+
+//EventMessage as a message
+type EventMessage struct {
+	ID string `json:"id"`
+}
+
+type RequestRequirement struct {
+	Url     string            `json:"url"`
+	Action  string            `json:"action"`
+	Payload []byte            `json:"payload"`
+	Header  map[string]string `json:"header"`
+	Key     string            `json:"key"`
+}
+
+//MaintenanceScheduler structure
+type MaintenanceScheduler struct {
+	rclt    *redisc
+	hkey    string
+	echan   chan error
+	schan   chan bool
+	debug   bool
+	handler SendRequestWithPubSub
+
+	sleepDuration time.Duration
+}
+
+//Configuration as maintenance scheduler preferences
+type Configuration struct {
+	RedisURL      string
+	ContexName    string
+	Debug         bool
+	SleepDuration time.Duration
+	Handler       SendRequestWithPubSub
+}
+
+//New creates new redis maintenance
+func New(config Configuration) (*MaintenanceScheduler, error) {
+	rclt, err := dial(config.RedisURL)
+	if err != nil {
+		return nil, err
+	}
+	return &MaintenanceScheduler{
+		rclt:          rclt,
+		hkey:          config.ContexName,
+		echan:         make(chan error, 1),
+		schan:         make(chan bool, 1),
+		debug:         config.Debug,
+		sleepDuration: config.SleepDuration,
+		handler:       config.Handler,
+	}, nil
+}
+
+//Run runs the maintainance scheduler
+func (m *MaintenanceScheduler) Run() {
+	rc := m.rclt.gconn()
+	psc := redis.PubSubConn{
+		Conn: rc,
+	}
+	key := fmt.Sprintf("%s", m.hkey)
+	if err := psc.PSubscribe(key); err != nil {
+		m.echan <- err
+		return
+	}
+	for {
+		select {
+		case <-m.schan:
+			err := rc.Close()
+			if err != nil {
+				log.Println("err", err)
+			}
+			err = psc.Close()
+			if err != nil {
+				log.Println("err", err)
+			}
+			return
+		default:
+			switch msg := psc.Receive().(type) {
+			case redis.Message:
+				go m.process(msg.Data)
```

Comment:
We should test and benchmark this. I believe it is from my code, but since we dont need to do the processes in parallel, can we not using go routine?